### DISCUSSION TITLE
NetBox status and configuration update from external system

### DISCUSSION THREAD

**Discussion by chaijunkin** [netbox-community/netbox#9357] [General]
Can I ask if I have a proxmox cluster (hypervisor) inventory in NetBox, it contains all the virtual machine. If a virtual machine shutdown gracefully or RAM is changed, how to update netbox virtual machine data in real time? 

Assuming the hypervisor, device and virtual machine platform has no webhook or the virtual machine is shutdown in the guest. How do you all update the netbox to represent the “Source of Truth”?  I attempt to figure out some alternatives but the solution seem didn’t change the root problem.

Solution 1: cron job on ansible to get ALL data from proxmox api and update in netbox (will this be dumb because it consumed ansible resource)
Solution 2: cron job on hypervisor + virtual machine, push event-based value to a webhook system to update netbox 
Solution 3: Is this should be done on the monitoring system, what if the monitoring system agent is stopped?
Solution 4: ???

Feel free to discuss ;( Idea is appreciated.

**Comment by mtinberg** [comment]
Question 0: if you aren't using the data in Netbox to push config why have it populated at all?  Source of Truth or Source of Intent means that the data you store in Netbox documents what the configuration should be that other config management tools like Ansible can use to effect configs.

I too have much data in Netbox that isn't a source of truth but is status information which is pulled from managed devices as Netbox can be a good place to aggregate, display and present that information to other systems.  In that case having an Ansible cron job that periodically audits and refreshes state into Netbox that Netbox isn't the authoritative source for seems fine, or a custom utility that runs over the API.  If having this data out of sync for a short time between when a change is made by and admin and when the Ansible audit cron job runs is a problem, then maybe there is more questions about why, but only then would it make sense to invest time in other more complex systems to update the data faster.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: chaijunkin ***@***.***>
Sent: Friday, May 13, 2022 6:48 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] NetBox status and configuration update from external system (Discussion #9357)


Can I ask if I have a proxmox cluster (hypervisor) inventory in NetBox, it contains all the virtual machine. If a virtual machine shutdown gracefully or RAM is changed, how to update netbox virtual machine data in real time?

Assuming the hypervisor, device and virtual machine platform has no webhook or the virtual machine is shutdown in the guest. How do you all update the netbox to represent the “Source of Truth”? I attempt to figure out some alternatives but the solution seem didn’t change the root problem.

Solution 1: cron job on ansible to get ALL data from proxmox api and update in netbox (will this be dumb because it consumed ansible resource)
Solution 2: cron job on hypervisor + virtual machine, push event-based value to a webhook system to update netbox
Solution 3: Is this should be done on the monitoring system, what if the monitoring system agent is stopped?
Solution 4: ???

Feel free to discuss ;( Idea is appreciated.

**Reply by chaijunkin** [reply]
Sorry for late reply. It makes sense to use netbox as ansible for configuration for Question 0 and this is what I’m doing. I did audit inventory every morning but this doesn’t avoid configuration conflict that you stated. I think building a complex and efficient system is overkilled. Would be good to stay with current status.

Maybe I make the wrong statement on source of truth and assume the inventory is represent the truth itself. Thank you for your explanation.
